network-area-diagram/src/main/java/com/powsybl/nad/build/iidm/VoltageLevelFilter.java

```diff
@@ -99,6 +101,71 @@ private static void traverseVoltageLevels(Set<VoltageLevel> voltageLevelsDepth,
         traverseVoltageLevels(nextDepthVoltageLevels, depth - 1, visitedVoltageLevels);
     }
 
+    private static void traverseVoltageLevelsWithVoltageFilter(Set<VoltageLevel> voltageLevelsDepth, int depth, Set<VoltageLevel> visitedVoltageLevels,
+                                                               double lowNominalVoltageBound, double highNominalVoltageBound) {
+        if (depth < 0) {
+            return;
+        }
+
+        Set<VoltageLevel> nextDepthVoltageLevels = new HashSet<>();
+        for (VoltageLevel vl : voltageLevelsDepth) {
+            if (!visitedVoltageLevels.contains(vl)
+                    && vl.getNominalV() >= lowNominalVoltageBound
+                    && vl.getNominalV() <= highNominalVoltageBound) {
+                traverseVoltageLevel(visitedVoltageLevels, nextDepthVoltageLevels, vl);
+            }
+        }
+        traverseVoltageLevelsWithVoltageFilter(nextDepthVoltageLevels, depth - 1, visitedVoltageLevels, lowNominalVoltageBound, highNominalVoltageBound);
+
+    }
+
+    private static void traverseVoltageLevel(Set<VoltageLevel> visitedVoltageLevels, Set<VoltageLevel> nextDepthVoltageLevels, VoltageLevel vl) {
+        visitedVoltageLevels.add(vl);
+        vl.visitEquipments(new VlVisitor(nextDepthVoltageLevels, visitedVoltageLevels));
+    }
+
+    public static VoltageLevelFilter createNominalVoltageFilter(Network network, List<String> voltageLevelIds,
+                                                                double lowNominalVoltageBound, double highNominalVoltageBound,
+                                                                int depth) {
+        Objects.requireNonNull(network);
+        Objects.requireNonNull(voltageLevelIds);
+        checkVoltageBoundValues(lowNominalVoltageBound, highNominalVoltageBound);
+        Set<VoltageLevel> startingSet = new HashSet<>();
+
+        for (String voltageLevelId : voltageLevelIds) {
+            VoltageLevel vl = network.getVoltageLevel(voltageLevelId);
+            if (vl == null) {
+                throw new PowsyblException(UNKNOWN_VOLTAGE_LEVEL + voltageLevelId + "'");
+            }
+            if (vl.getNominalV() < lowNominalVoltageBound || vl.getNominalV() > highNominalVoltageBound) {
+                throw new PowsyblException("vl '" + voltageLevelId +
```

Comment:
What I did is the following : when the voltage level is out of bounds, it is not added to the `startingSet`.

@flo-dup, you may want it to be added to the `startingSet` anyway but with a warning issued. Let me know, I can change that ! We only have to be careful about the fact that it will change the bahavior Python users were used to (but it may go unnoticed :D)

Reply:
I'm not 100% convinced by any of the two. I'd say it's a bit better to include it in the startingSet. If the user doesn't want to see that voltage level, he can filter it out before.